src/core/protocol-authorization.ts

```diff
@@ -274,6 +275,17 @@ export class ProtocolAuthorization {
 
     const actionRules = inboundMessageRuleSet.$actions;
     if (author === tenant) {
+      // tenant is always authorized
+      return;
+    } else if (incomingMessage.author !== undefined && incomingMessage.authorizationPayload?.permissionsGrantId !== undefined) {
+      // PermissionsGrant gives the author explicit access to this record
+      await RecordsGrantAuthorization.authorizeRecordsGrant(
+        tenant,
+        incomingMessage,
+        recordsWrite,
+        incomingMessage.author,
+        messageStore
+      );
```

Comment:
Couple of questions for my understanding:
1. Why would a record written using a permission grant end up in this code path? I guess another way to ask the question: I saw in code further down that you flip the order or grant authorization and protocol authorization, what's the trigger of that?
1. Does a permission grant not have the capability to grant either read or read-write capability? If not, then it seems like we ought to have it for what seems to be very common scenarios.

Reply:
1. The context is that `ProtocolAuthorization` does more than just the actual authorization of `RecordsWrites`. It also validates the structure of the `RecordsWrites`, e.g. validating the `protocolPath`. So, protocol `RecordsWrites` must enter `ProtocolAuthorization` even if a `PermissionsGrant` exists. After the previous steps of `ProtocolsAuthorization` are done, then actual authorization happens where we verify the grant.

2. A grant can be for `RecordsRead` or for `RecordsWrite`, but not both. Two separate grants are needed in order to give read AND write access.

Reply:
@diehuxx got it, thanks! So the order is critical,  must use `ProtocolAuthorization.authorize()` whenever `protocol` exists, even if a `permissionsGrantId` also exist.

Reply:
Added a comment